### ISSUE TITLE
AppleClang identification dependes on CMP0025

### ISSUE THREAD

**Issue by kolis** [conan-io/cmake-conan#33]
Identification of `AppleClang` compiler in `CMAKE_CXX_COMPILER_ID` depends on [CMP0025](https://cmake.org/cmake/help/v3.0/policy/CMP0025.html) being set to `NEW`.

However the default behavior is `OLD` 
```
Unlike most policies, CMake version 3.0.2 does not warn by default when this policy is not set and simply uses OLD behavior. See documentation of the CMAKE_POLICY_WARNING_CMP0025 variable to control the warning.
```
To help detect this issue `cmake-conan` should enable the `CMAKE_POLICY_WARNING_CMP0025` flag and/or check the value of `CMP0025` policy on Mac builds. Also note that this policy must be set before the `project` declaration.

Otherwise we just get this error
```
ERROR: Invalid setting '7.0' is not a valid 'settings.compiler.version' value.
Possible values are ['3.3', '3.4', '3.5', '3.6', '3.7', '3.8', '3.9']
```
because it is identified as `Clang` instead of `AppleClang`

**Comment by memsharded** [comment]
Thanks for reporting this. Just a question, this happens irrespective of the CMake version? Or the policy must be conditioned to such version? I'd say that this worked at some point for OSX and AppleClang.

**Comment by kolis** [comment]
I don't know, I just recently starting using conan. I have only tested with my current cmake version 3.7.2.

**Comment by lasote** [comment]
Working with the policies should work irrespective of the CMake version, at least afaik.

**Comment by lasote** [comment]
Ok, I think I get it. Actually, CMake is detecting "Clang", that matches with the CMake documentation. 

The thing is, ``conanbuildinfo.cmake`` doesn't care, the actions done by clang and apple-clang are the same and the compiler check won't complain if settings are apple-clang and CMake detects clang.

So, What compiler do you have in ``conan.conf``, probably is ``clang`` when it should be ``apple-clang``, actually the versions check is performed by conan, not by CMake. 

Could you confirm this?
Thanks

**Comment by kolis** [comment]
Hmm, nope, sorry, my `conan.conf` looks like this:
```
[settings_defaults]
arch=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=7.0
os=Macos
```

It is definitely affected by CMP0025 for me. If I set `cmake_policy(SET CMP0025 OLD)` i get
```
-- Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
-- Conan executing: conan install -f=/Volumes/Big/Work/agx/branch/conanfile.txt -g cmake -s build_type=Debug -s os=Macos -s compiler=clang -s compiler.version=7.0 -s compiler.libcxx=libstdc++ --build=missing
ERROR: Invalid setting '7.0' is not a valid 'settings.compiler.version' value.
Possible values are ['3.3', '3.4', '3.5', '3.6', '3.7', '3.8', '3.9', '4.0']
```
And printing `CMAKE_CXX_COMPILER_ID` gives `Clang`.

And with `cmake_policy(SET CMP0025 NEW)` I get
```
CMAKE_CXX_COMPILER_ID: AppleClang
-- Conan ** WARNING** : This detection of settings from cmake is experimental and incomplete. Please check 'conan.cmake' and contribute
-- Conan executing: conan install -f=/Volumes/Big/Work/agx/branch/conanfile.txt -g cmake -s build_type=Debug -s os=Macos -s compiler=apple-clang -s compiler.version=7.0 -s compiler.libcxx=libc++ --build=missing
```

And printing `CMAKE_CXX_COMPILER_ID` gives `AppleClang`.



In `conan.cmake` there is this test of `CMAKE_CXX_COMPILER_ID`
```
elseif (${CMAKE_CXX_COMPILER_ID} STREQUAL AppleClang)
```

**Comment by memsharded** [comment]
Hi @kolis 

I see, thanks! The only issue is that I think that it is not possible for ``conan.cmake`` to set it. Note how it is used:
```cmake
cmake_minimum_required(VERSION 2.8)
project(conan_wrapper CXX)
include(conan.cmake)
```

but as you said, and the docs clearly state, the policy MUST be set before the project declaration:

```cmake
cmake_policy(SET CMP0025 NEW)
cmake_minimum_required(VERSION 2.8)
project(conan_wrapper CXX)
include(conan.cmake)
```
There is nothing in the ``conan.cmake`` file we provide that could make cmake understand that it is the AppleClang compiler. I guess the only way is to document it, any other idea or suggestion? 

As a side note, I have to admit that I am not a big fan of this cmake-conan project, because of this kind of issues, I think it is much better to call conan directly, or to have python scripts that orchestrate the build, calling the necessary tools (conan, cmake, or others), much better scripting in python than having to script in cmake. Just my personal opinion.

**Comment by memsharded** [comment]
@lasote just realized, there might be a workaround. We might try to assume ``apple-clang`` if we are in apple and the compiler is ``clang``, and the policy is not set. For those rare cases using bare ``clang`` in apple, they should set the policy to disable the assumption. 
I think this should cover well the default case, and leave a way to define it for special cases. Will try to submit a fix.

**Comment by memsharded** [comment]
Done it. 
I have also added CI with travis-ci for both Linux and OSX, and it seems the approach is running fine. A message outputs:

```
- Conan: APPLE and Clang detected. Assuming apple-clang compiler. Set CMP0025 to avoid it
```

It is already merged in master, I am going to tag this too. Please try it and report, re-open issue if something fails, or have any question. Thanks!

**Comment by memsharded** [comment]
Included also in tagged 0.4 pre-release
